smarts/core/smarts.py

```diff
@@ -905,6 +906,7 @@ def _try_emit_envision_state(self, provider_state, obs, scores):
             speed=speed,
             heading=heading,
             lane_ids=lane_ids,
+            frame_time=time(),
```

Comment:
(Sorry, this is my fault for being too confusing yesterday when we talked.  I know I showed it like this then at one point.)

Using `time()` here should help ensure that the relative timing between frames played by Envision is the same as how they were produced.  But since we currently don't produce them at a consistent rate in Smarts (because there's no explicit delay in our step loop), the frame rate will not line up with Smarts' nominal `timestep_sec`.

There are a few options here.  

The _easiest_ would be to continue to rely on the wait around line 236 of `envision/server.py` to enforce that frames cannot be replayed too quickly in the web client.  (But doing this still requires including the _simultation_ time to the web client, not `time()` as it is now... see below.)  However, although it might be an adequate short-term solution, I think this is sub-optimal, as I tried to explain yesterday.  For example, that does not allow for a non-fixed-rate clock, if/when we ever want one (which I think we might at some point).  More importantly, it doesn't allow for the web client itself to make any intelligent decisions about how it might want to buffer or save frames, depending on what the user is doing (cf. the example we discussed of "PVR-ing a simulation").

Anyway, for fixed frame rates, we need to send the _simulation_ time (i.e.,always a multiple of `timestep_sec`) from Smarts.  This will require updating `simulation.js` too (as it currently considers the frame time to be a real timestamp).  

Also as mentioned above, we _might_ _also_ want to allow an option for the `dt` in the step loop to reflect real time (i.e., a variable frame rate), an example of which I showed you yesterday.  But I think there will probably always be other users that would still prefer a fixed frame rate too, so this would have to be done as an optional thing.  If we did this, in this mode we would still send the simulation time (which would now more accurately track real time but would start at 0 instead of seconds from the epoch).  The way `simulation.js` is _currently_ written would probably work ok in this case.

The risk with the fixed-step approaches is that Envision will fall behind Smarts, as Envision's frames may be playing at a "true" 10 Hz while Smarts is producing them much faster, and/or frame buffers will overflow.  So my suggestion would be to do this _in conjunction with_ the change I suggested in PR #741, which would throttle the Smarts' clock when running in (near) real-time mode.  @Gamenot may have more thoughts on this though.

To summarize:
- For Envision playback to be "near real time", the relative gap between played frames should match the (nominal) gap between frames in Smarts (which right now is fixed at 0.1 exactly).
    - Doing this (without relaying on line 236 of `server.py`) will require sending the simulation time from Smarts
    - Doing this _may_ require either throttling the step rate in Smarts such that the real time matches the nominal time (when not running headless as least), as suggested in PR #741, OR being careful about Frame buffering on the Envision side.
- We _may_ _also_ want to add a fixed vs. variable rate flag for the Smarts step loop (where in the variable rate case, `dt` will be based on the _actual_ real time elapsed per step loop).  
    - This would require line 236 of `server.py` to be changed (and if so, we would want to avoid re-sending the same frame multiple times).

Reply:
My preference here is to have no limitation on the SMARTS step and instead impose the limitations on Envision. I feel that limitations should likely be applied and mostly configurable via the Envision web client since we should be treating Envision as a playback tool similar to a video player. 

The main two useful options are:
- Uncapped - play frames as fast as possible (as fast as they are received)
- Approximate realtime - play simulation frames as if it were in realtime (i.e. play simulation time frames between the 9s and 10s timestamps over 1 second realtime) 

If we are running the playback at realtime we should not worry about the playback falling behind.

Reply:
@sah-huawei Thanks for the detailed consideration! The following are what I understood from this and what I will do:

- we want to set the `frame_time` to be something like: "0, 1*timestep_sec, 2*timestep_sec, 3*timestep_sec.....", this will be what we define as the real time of the frame, needs to change `simulation.js`
- Make it configurable on Envision side for 2 options as @Gamenot mentioned:
"uncapped" would disregard the `frame_time` info and "play as fast as frames are received.(needs to change `simulation.js` to not sleep according to frame_time on each frame, but according to a small fixed amount enough for React to trigger update)
"Approximate realtime" would play the frames according to the `frame_time` on each frame. (this is what current simulation.js supports, since it plays each frame according to its frame_time, so updating frame_time on smarts should be enough)

Is my understanding correct? Please correct me if I am wrong. Otherwise I will start working towards this

Reply:
sounds good! yes, I was also planning to change the sleep behaviour in `server.py` for the uncapped mode. Switching on the uncapped mode would send a http request from client to server and server will have a flag to switch between the 2 behaviours ("uncapped" vs "Approximate realtime").